core/src/main/scala/org/ensime/core/Indexer.scala

```diff
@@ -24,10 +24,26 @@ class Indexer(
     LineSourcePositionHelper.fromFqnSymbol(hit)(vfs)
   )
 
-  def oldSearchTypes(query: String, max: Int) =
-    index.searchClasses(query, max).filterNot {
-      name => name.fqn.endsWith("$") || name.fqn.endsWith("$class")
-    }.map(typeResult)
+  def oldSearchTypes(query: String, max: Int) = {
+    val scalaNames = scala.collection.mutable.Set.empty[String]
+
+    def isUnique(fqn: String): Boolean = {
+      val stripped =
+        if (fqn.endsWith("$")) fqn.dropRight(1)
+        else if (fqn.endsWith("$class")) fqn.dropRight(6)
+        else fqn
+
+      if (scalaNames.contains(stripped)) {
+        false
+      } else {
+        scalaNames.add(stripped)
+        true
+      }
+    }
+
+    index.searchClasses(query, max)
```

Comment:
Guys, you think much faster than me, let's slow down for a moment :D 
I do not get what you want to do in first `map`. I get `FqnSymbol` and need to convert it to `ClassDef` with changed `ClassDef#fqn`. So basically I would remap trait to case class in this case. Would you really implement it like that?